src/chef-server-ctl/plugins/maintenance.rb

```diff
@@ -0,0 +1,97 @@
+
+require "redis"
+require "resolv"
+
+def redis
+  @redis ||= begin
+               vip = running_config["private_chef"]["redis_lb"]["vip"]
+               port = running_config["private_chef"]["redis_lb"]["port"]
+               password = credentials.get('redis_lb', 'password')
+               Redis.new(port: port, ip: vip, password: password)
+             end
+end
+
+def enable_maintenance
+  puts "- Enabling maintenance mode"
+  redis.hset("dl_default", "503_mode", true)
+end
+
+def disable_maintenance
+  puts "- Disabling maintenance mode"
+  redis.hdel("dl_default", "503_mode")
+end
+
+def list_allowed_ips()
+  ips = redis.smembers("xmaint_allowed_ips_list")
+  if ips.length == 0
+    puts "- allowed IP list is empty."
+  else
+    puts "- The following IPs are allowed in the maintenance mode \n", ips
+  end
+end
+
+def add_ip_to_allowed_list(ip)
+  validate_ip(ip)
+  puts "- Adding #{ip} to allowed list."
+  redis.sadd("xmaint_allowed_ips_list", ip)
+end
+
+def remove_ip_from_allowed_list(ip)
+  validate_ip(ip)
+  puts "- Removing #{ip} from allowed list."
+  redis.srem("xmaint_allowed_ips_list", ip)
+end
+
+def validate_ip(ip)
+  isIPv4 = ip =~ Resolv::IPv4::Regex ? true : false
+  isIPv6 = ip =~ Resolv::IPv6::Regex ? true : false
+  unless isIPv4 or isIPv6
+    $stderr.puts "The IP address is invalid, please enter valid IP address"
+    exit 1
+  end
+end
+
+
+add_command_under_category "maintenance", "general", "Handel the server properly while maintenance", 2 do
+  args = ARGV[1..-1] # Chop off first 1 args, keep the rest... that is, everything after "chef-server-ctl maintenance"
+  options = {}
+
+  case args[0]
+    when 'on'
+      enable_maintenance
+      args = args[1..-1]
+    when 'off'
+      disable_maintenance
+      args = args[1..-1]
+    else
+      if args.length == 0
+        $stderr.puts "Please specify the action 'on' or 'off' or pass the maintenance options"
```

Comment:
```
root@api:~# chef-server-ctl maintenance x
root@api:~#
root@api:~# chef-server-ctl maintenance blah
root@api:~#
```
Can we catch errors of this type?

Reply:
Thanks for the Inputs! I have updated the code here - https://github.com/chef/chef-server/blob/09df8902773c10fdb778cd6aace913741b58d813/src/chef-server-ctl/plugins/maintenance.rb#L55

Reply:
Nice.  Would you like to paste the results of testing the updated code here?